### PR TITLE
uadk: fix some uadk framework issues

### PR DESCRIPTION
**Pull request by Liulongfang**
According to the needs of device driver registration. Algorithm-driven
priorities and types need to be distinguished.

In addition, the loading function of the device driver has been fixed.

### CODE REVIEW

**Comment by gaozhangfei** [comment]
wd_xxx_uninit2 should not check xx_setting.priv, but using  wd_alg_get_init to prevent re-enter, status is key to prevent not priv.

**Comment by gaozhangfei** [comment]
and which func in uninit can not re-enter?

**Comment by gaozhangfei** [comment]
UADK_ALG_HW/SW is type or priority, enum alg_priority?
can drv define priority in any number?  no max, min etc.

**Comment by Liulongfang** [comment]
> wd_xxx_uninit2 should not check xx_setting.priv, but using wd_alg_get_init to prevent re-enter, status is key to prevent not priv.


This method is possible before modifying the priv allocation of the setting of the API

**Comment by Liulongfang** [comment]
> and which func in uninit can not re-enter?

Resource allocations involving global settings cannot be re-entered

**Comment by Liulongfang** [comment]
> UADK_ALG_HW/SW is type or priority, enum alg_priority? can drv define priority in any number? no max, min etc.

UADK_ALG_HW/SW indicates the type of driver. enum alg_priority?
is the algorithm priority. Consistent with the meaning of the kernel's algorithm priority, the minimum value is 0, and the maximum value has no upper limit

**Comment by gaozhangfei** [comment]
> > and which func in uninit can not re-enter?
> 
> Resource allocations involving global settings cannot be re-entered

can you clarify which func name? why not make it re-enter?

I don't think check priv can prevent re-enter, 
for example, two threads enter uninit before priv is free, then thread1 free priv, thread2 continue do uninit.

**Comment by gaozhangfei** [comment]
> > UADK_ALG_HW/SW is type or priority, enum alg_priority? can drv define priority in any number? no max, min etc.
> 
> UADK_ALG_HW/SW indicates the type of driver. enum alg_priority? is the algorithm priority. Consistent with the meaning of the kernel's algorithm priority, the minimum value is 0, and the maximum value has no upper limit

ok, so change enum alg_priority -> enum alg_type?

enum alg_priority {
        UADK_ALG_SOFT = 0x0,
        UADK_ALG_CE_INSTR = 0x1,
        UADK_ALG_SVE_INSTR = 0x2,
        UADK_ALG_HW = 0x3
};

**Comment by Liulongfang** [comment]
> 

Yes, the name of the enumeration class should be changed here.

**Comment by Liulongfang** [comment]
> 

Yes, the name of the enumeration class should be changed here.

**Comment by gaozhangfei** [comment]
wd_alg_check_available
case UADK_ALG_SOFT
return true?

**Comment by Liulongfang** [comment]
> wd_alg_check_available case UADK_ALG_SOFT return true?

No, There is no soft calculation yet, it returns false

**Comment by gaozhangfei** [comment]
> > wd_alg_check_available case UADK_ALG_SOFT return true?
> 
> No, There is no soft calculation yet, it returns false

soft is CPU, always exists
If I add zlib, need modify here to register?

**Comment by Liulongfang** [comment]
> > > wd_alg_check_available case UADK_ALG_SOFT return true?
> > 
> > 
> > No, There is no soft calculation yet, it returns false
> 
> soft is CPU, always exists If I add zlib, need modify here to register?

Yes, if you add CPU soft calculation, here you need to add CPU soft calculation check, and return true or false according to the actual result

**Comment by gaozhangfei** [comment]
> > > wd_alg_check_available case UADK_ALG_SOFT return true?
> > 
> > 
> > No, There is no soft calculation yet, it returns false
> 
> soft is CPU, always exists If I add zlib, need modify here to register?

and is there any method to judge whether system support sve or ce?

> > > > wd_alg_check_available case UADK_ALG_SOFT return true?
> > > 
> > > 
> > > No, There is no soft calculation yet, it returns false
> > 
> > 
> > soft is CPU, always exists If I add zlib, need modify here to register?
> 
> Yes, if you add CPU soft calculation, here you need to add CPU soft calculation check, and return true or false according to the actual result

then what's the point of this check? to prevent other drivers from registering to uadk?

**Comment by Liulongfang** [comment]
> 
Yes, the software can check whether the current platform supports SVE or CE by obtaining the instruction set supported by the CPU。


Used to prevent unsupported algorithms from being registered on alg_list. Otherwise, useless alg will be mounted on this list, which will affect the driver query.

**Review by haofang111 — COMMENTED** [review]
ok

File: drv/hisi_hpre.c [hunk 1]
```diff
@@ -2487,37 +2497,41 @@ static struct wd_alg_driver hpre_dh_driver = {
...
 	ret = wd_alg_driver_register(&hpre_rsa_driver);
-	if (ret)
+	if (ret && ret != -WD_ENODEV)
 		WD_ERR("failed to register HPRE rsa driver!\n");
 
 	ret = wd_alg_driver_register(&hpre_dh_driver);
-	if (ret)
-		WD_ERR("failed to register HPRE dh driver!\n");
+	if (ret && ret != -WD_ENODEV)
+		WD_ERR("failed to register HPRE rsa driver!\n");
```
[the remark this answers is not in the document]
Comment (Liulongfang) [reply drv/hisi_hpre.c:2516]: OK

[the remark this answers is not in the document]
Comment (haofang111) [reply drv/hisi_hpre.c]: /LGTM

**Pull request merged**